stats/opentelemetry/e2e_test.go

```diff
@@ -143,102 +143,662 @@ func setupStubServer(t *testing.T, metricsOptions *opentelemetry.MetricsOptions,
 	return ss
 }
 
-// TestMethodAttributeFilter tests the method attribute filter. The method
-// filter set should bucket the grpc.method attribute into "other" if the method
-// attribute filter specifies.
-func (s) TestMethodAttributeFilter(t *testing.T) {
-	maf := func(str string) bool {
-		// Will allow duplex/any other type of RPC.
-		return str != testgrpc.TestService_UnaryCall_FullMethodName
-	}
-	mo, reader := defaultMetricsOptions(t, maf)
-	ss := setupStubServer(t, mo, nil)
-	defer ss.Stop()
-
-	// Make a Unary and Streaming RPC. The Unary RPC should be filtered by the
-	// method attribute filter, and the Full Duplex (Streaming) RPC should not.
-	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
-	defer cancel()
-	if _, err := ss.Client.UnaryCall(ctx, &testpb.SimpleRequest{Payload: &testpb.Payload{
-		Body: make([]byte, 10000),
-	}}); err != nil {
-		t.Fatalf("Unexpected error from UnaryCall: %v", err)
-	}
-	stream, err := ss.Client.FullDuplexCall(ctx)
-	if err != nil {
-		t.Fatalf("ss.Client.FullDuplexCall failed: %f", err)
-	}
+// waitForTraceSpans waits until the in-memory span exporter has received the
+// expected trace spans based on span name and kind. It polls the exporter at a
+// short interval until the desired spans are available or the context is
+// cancelled.
+//
+// Returns the collected spans or an error if the context deadline is exceeded
+// before the expected spans are exported.
+func waitForTraceSpans(ctx context.Context, exporter *tracetest.InMemoryExporter, wantSpans []traceSpanInfo) (tracetest.SpanStubs, error) {
+	for ; ctx.Err() == nil; <-time.After(time.Millisecond) {
+		spans := exporter.GetSpans()
+		allFound := true
+
+		for _, wantSpan := range wantSpans {
+			found := false
+			for _, span := range spans {
+				if span.Name == wantSpan.name && span.SpanKind.String() == wantSpan.spanKind {
+					found = true
+					break
+				}
+			}
+			if !found {
+				allFound = false
+				break
+			}
+		}
 
-	stream.CloseSend()
-	if _, err = stream.Recv(); err != io.EOF {
-		t.Fatalf("stream.Recv received an unexpected error: %v, expected an EOF error", err)
-	}
-	rm := &metricdata.ResourceMetrics{}
-	reader.Collect(ctx, rm)
-	gotMetrics := map[string]metricdata.Metrics{}
-	for _, sm := range rm.ScopeMetrics {
-		for _, m := range sm.Metrics {
-			gotMetrics[m.Name] = m
+		if allFound {
+			return spans, nil
 		}
 	}
+	if ctx.Err() != nil {
+		return nil, fmt.Errorf("error waiting for complete trace spans %v: %v", wantSpans, ctx.Err())
+	}
+	return exporter.GetSpans(), nil
+}
 
-	wantMetrics := []metricdata.Metrics{
+// traceDataWithCompressor returns a traceSpanInfo for a unary RPC and
+// streaming RPC with certain compression and message flow sent, when
+// compressor is used
+func traceDataWithCompressor() []traceSpanInfo {
```

Comment:
Can this be inlined, like how it was before? The input and output are important parts of the test that should not be abstracted.